tools/python/mwm_downloader.py

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import re
+import time
+import sys
+import urllib.request
+import urllib.parse
+import urllib.error
+import logging
+
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+
+DIRECT_MAP_URL = 'http://direct.mapswithme.com/direct/'
+ROOT = os.path.dirname(os.path.realpath(__file__))
+DEFAULT_DOWNLOAD_DIRECTORY = os.path.join(os.path.dirname(os.path.dirname(ROOT)), 'data')
+MAP_REVISION_REGEXP = re.compile(r'(\d{6})')
+MWM_NAME_REGEXP = re.compile(r'"(\S+\.mwm)"')
+
+logger = logging.getLogger(__name__)
+
+
+def parse_html_by_regexp(url, regexp):
+    def parse_line(line):
+        return regexp.search(line.decode('utf-8'))
+
+    with urllib.request.urlopen(url) as f:
+        lines = f.readlines()
```

Comment:
You do not need readlines if urlopen returns a file-like object

Reply:
This object supports iterating line by line, but I've finally returned to generator-style implementation, because this version is not working:
```python
with urllib.request.urlopen(url) as lines:
    filtered_lines = filter(None, map(parse_line, lines))
    return (line.group(1) for line in filtered_lines)
```
No idea why.

Reply:
Looks like the problem is in ```urllib.request.urlopen``` context manager. Connection is closing without reading data, when you exit from it with  ```return``` statement. Take a look on this little example:
```python

class foo_context_manager:
    def __init__(self):
        self.bar = 10

    def __iter__(self):
        while self.bar:
            yield self.bar
            self.bar -= 1

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.bar = 0


def foo():
    with foo_context_manager() as iterator:
        return map(lambda n: n ** 2, iterator)


def foo_yield():
    with foo_context_manager() as iterator:
        for i in iterator:
            yield i ** 2

if __name__ == '__main__':

    print('Using map function:')
    for bar in foo():
        print(bar)

    print('\n\n')

    print('Using yield:')
    for bar in foo_yield():
        print(bar)
```
Output is:
```
Using map function:



Using yield:
100
81
64
49
36
25
16
9
4
1
```

Reply:
Mb

for line in filtered_lines:
yield line.group(1)

then?

Reply:
Fixed!